Introduce Ml Inference Search Request Extension

mingshl opened the pull request:
### Description
Introduce Ml Inference Search Request Extension, this is adding ml_inference search extension during search request/query phase,

For the search response side, I already introduced a  `ml_inference` search response extension in earlier PR released in 2.18.  [Support ML Inference Search Processor Writing to Search Extension](https://github.com/opensearch-project/ml-commons/commit/09aa6ea0b108c29ff074292278370c2f5e772688)

### Related Issues
 https://github.com/opensearch-project/ml-commons/issues/3286
https://github.com/opensearch-project/ml-commons/issues/3054

### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dhrubo-os commented:
@mingshl let's take the update from main. File has conflict.

mingshl commented:
thanks @dhrubo-os ， rebased and resolved conflicts.

dhrubo-os commented:
Tests are failing

mingshl commented:
> Tests are failing

there is no changes related to ConversationalMemory, will try rerun the tests. 
```
> Task :opensearch-ml-memory:test

ConversationalMemoryHandlerITTests > testDifferentUsers_DifferentConversations FAILED
    java.lang.NoClassDefFoundError at ConversationalMemoryHandlerITTests.java:389
        Caused by: java.lang.ClassNotFoundException at ConversationalMemoryHandlerITTests.java:389

```

mingshl commented:
rebased and resolved conflicts.

dhrubo-os commented:
```
D:\a\ml-commons\ml-commons\common\build\generated\sources\delombok\java\main\org\opensearch\ml\common\utils\StringUtils.java:379: error: unknown tag: implNote
     * @implNote This method uses JsonPath for JSON manipulation and StringUtils for path existence checks.
       ^
```

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:
```diff
@@ -273,13 +280,52 @@ public void onFailure(Exception e) {
-                    Object modelOutputValue = getModelOutputValue(mlOutput, modelOutputFieldName, ignoreMissing, fullResponsePath);
-                    String jsonPathExpression = "$." + newQueryField;
-                    JsonPath.parse(incomeQueryObject).set(jsonPathExpression, modelOutputValue);
+                    String newQueryField = null;
```
this declaration can be moved inside the try/catch block

mingshl replied:
if moved inside the catch block then when throw exception, we cannot get the queryField information in the exception message

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:
```diff
@@ -273,13 +280,52 @@ public void onFailure(Exception e) {
+                        requestContext.setAttribute(newQueryField, modelOutputValue);
+
+                    } catch (PathNotFoundException e) {
+                        throw new IllegalArgumentException("can not find path " + newQueryField + "in query string");
```
do you want to log the full exception message here?

mingshl replied:
added logs

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:
```diff
@@ -358,11 +404,12 @@ private boolean validateQueryFieldInQueryString(
+                    if (!queryField.startsWith("ext.") && !queryField.startsWith("$.ext.")) {
+                        Object pathData = jsonData.read(queryField);
+                        if (pathData == null) {
+                            throw new IllegalArgumentException();
```
please add more information so user can have more information on the cause and address the issue

mingshl replied:
added error message

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java:
```diff
@@ -334,4 +341,26 @@ default List<String> writeNewDotPathForNestedObject(Object json, String dotPath)
         return path.replaceAll("\\[(\\d+)\\]", "$1\\.").replaceAll("\\['(.*?)']", "$1\\.").replaceAll("^\\$", "").replaceAll("\\.$", "");
     }
+
+    default void setRequestContextFromExt(SearchRequest request, PipelineProcessingContext requestContext) {
```
this class looks more like an abstract class rather then an interface, there are verbose default implementations for every method. Consider refactoring interface->abstract class sometime in the future

mingshl replied:
will take down this note for future refactoring in a separate PR.

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java:
```diff
@@ -334,4 +341,26 @@ default List<String> writeNewDotPathForNestedObject(Object json, String dotPath)
+                mlParams
+                    .forEach(
+                        (key, value) -> requestContext
+                            .setAttribute(String.format("ext.%s.%s", MLInferenceRequestParametersExtBuilder.NAME, key), value)
```
you need to use locale with `String.format`, default should be good enough

mingshl replied:
agree

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java:
```diff
@@ -334,4 +341,26 @@ default List<String> writeNewDotPathForNestedObject(Object json, String dotPath)
+            if (ext instanceof GenerativeQAParamExtBuilder) {
+                GenerativeQAParamExtBuilder qaParamExtBuilder = (GenerativeQAParamExtBuilder) ext;
+                Map<String, Object> mlParams = (Map<String, Object>) qaParamExtBuilder.getParams();
+                mlParams.forEach((key, value) -> requestContext.setAttribute(String.format("ext.%s.%s", PARAMETER_NAME, key), value));
```
same comment regarding the local

mingshl replied:
added `Locale.ROOT`

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParameters.java:
```diff
@@ -0,0 +1,79 @@
+    static final String ML_INFERENCE_FIELD = "ml_inference";
+
+    @Setter
+    @Getter
```
you can use @Data, it will do same as Setter+Getter

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParameters.java:
```diff
@@ -0,0 +1,79 @@
+    @Getter
+    private Map<String, Object> params;
+
+    public MLInferenceRequestParameters(Map<String, Object> params) {
```
if you add `final` to `param` field you can replace this constructor with Lombok's AllArgsContructor

mingshl replied:
later during search processors, possibly, this params is going to get added with new value, so trying not to make it final

dhrubo-os replied:
The final keyword ensures that the reference to the map cannot be changed to point to another map, but it does not make the contents of the map immutable. You can still modify the contents, such as adding, updating, or removing items from the map.

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParametersExtBuilder.java:
```diff
@@ -0,0 +1,82 @@
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.search.SearchExtBuilder;
+
+public class MLInferenceRequestParametersExtBuilder extends SearchExtBuilder {
```
Please add class level java doc that describes class responsibilities

mingshl replied:
added

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParametersExtBuilder.java:
```diff
@@ -0,0 +1,82 @@
+
+public class MLInferenceRequestParametersExtBuilder extends SearchExtBuilder {
+    private static final Logger logger = LogManager.getLogger(MLInferenceRequestParametersExtBuilder.class);
+    public static final String NAME = ML_INFERENCE_FIELD;
```
you can use ML_INFERENCE_FIELD directly, NAME is not needed

mingshl replied:
I was following other search extension format, all have a .NAME so trying to make it follow the same format

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParametersUtil.java:
```diff
@@ -0,0 +1,38 @@
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.search.SearchExtBuilder;
+
+public class MLInferenceRequestParametersUtil {
```
please add class level comments

mingshl replied:
added

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParametersUtil.java:
```diff
@@ -0,0 +1,38 @@
+                .collect(Collectors.toList());
+
+            if (!extBuilders.isEmpty()) {
+                mLInferenceRequestParametersExtBuilder = (MLInferenceRequestParametersExtBuilder) extBuilders.get(0);
```
if scenario with more than one element is valid did we document current behavior of using only first element and ignoring everything else? Btw, is this guaranteed that ext elements in this list will be in the same order as in user requst?

mingshl replied:
the extBuilders is filter by the name "ml_inference", so there will be the only type `MLInferenceRequestParametersExtBuilder`, and we are only getting the first MLInferenceRequestParametersExtBuilder. The worst case, if users use two MLInferenceRequestParametersExtBuilders, the second one is skip. We want to have one MLInferenceRequestParametersExtBuilder in every request. Hope this make sense

martin-gaievski commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParametersUtil.java:
```diff
@@ -0,0 +1,38 @@
+                .filter(extBuilder -> MLInferenceRequestParametersExtBuilder.NAME.equals(extBuilder.getWriteableName()))
+                .collect(Collectors.toList());
+
+            if (!extBuilders.isEmpty()) {
```
I suggest we check for `extBuilders.isEmpty()` condition first and fail fast by returning `null`. This way we can organize code on lines 32-36

mingshl replied:
make sense.

austintlee commented:
Looks good overall, but can you add REST test cases with some non trivial examples?

dhrubo-os commented on common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:
```diff
@@ -347,6 +349,114 @@ public static JsonObject getJsonObjectFromString(String jsonString) {
+     *             JsonPath expression (e.g., "$.store.book[0].title").
+     * @return true if the path exists in the JSON object, false otherwise.
+     * @throws IllegalArgumentException if the json object is null or if the path is null or empty.
+     * @throws PathNotFoundException if there's an error in parsing the JSON or the path.
```
we don't throw this exception, right? We catch it and then return false. Can we update the doc?

mingshl replied:
updated, thanks!!

dhrubo-os commented on memory/build.gradle:
```diff
@@ -41,6 +41,7 @@ dependencies {
     testImplementation("com.fasterxml.jackson.core:jackson-annotations:${versions.jackson}")
     testImplementation("com.fasterxml.jackson.core:jackson-databind:${versions.jackson_databind}")
     testImplementation group: 'com.networknt' , name: 'json-schema-validator', version: '1.4.0'
+    testImplementation 'com.jayway.jsonpath:json-path:2.9.0'
```
Just curious why do we need this here? I don't see any tests in memory module.

mingshl replied:
if I remove this dependency, it's throwing error in the build
```
ConversationMetaIndexITTests > testConversationsCanBeDeleted FAILED
    java.lang.NoClassDefFoundError at ConversationMetaIndexITTests.java:268
        Caused by: java.lang.ExceptionInInitializerError at IndexUtils.java:126

ConversationMetaIndexITTests > testCanQueryOverConversations FAILED
    java.lang.NoClassDefFoundError at ConversationMetaIndexITTests.java:431
        Caused by: java.lang.ExceptionInInitializerError at IndexUtils.java:126

ConversationMetaIndexITTests > testCanGetAConversationById FAILED
    java.lang.NoClassDefFoundError at ConversationMetaIndexITTests.java:539
        Caused by: java.lang.ExceptionInInitializerError at IndexUtils.java:126

ConversationMetaIndexITTests > testCanQueryOverConversationsByAdditionalInfo FAILED
    java.lang.NoClassDefFoundError at ConversationMetaIndexITTests.java:679
        Caused by: java.lang.ExceptionInInitializerError at IndexUtils.java:126

ConversationMetaIndexITTests > testCanAddNewConversation FAILED
    java.lang.NoClassDefFoundError at ConversationMetaIndexITTests.java:152
        Caused by: java.lang.ExceptionInInitializerError at IndexUtils.java:126
```

dhrubo-os commented on plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:
```diff
@@ -219,13 +226,15 @@ private void rewriteQueryString(SearchRequest request, String queryString, Actio
      * @param queryString      the original query string
      * @param requestListener  the {@link ActionListener} to be notified when the query string or query template is updated
      * @param processOutputMap the list of output mappings
+     * @param requestContext
```
add a description?

mingshl replied:
added and rebased

dhrubo-os commented on plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchRequestProcessor.java:
```diff
@@ -273,13 +280,53 @@ public void onFailure(Exception e) {
+                        Object modelOutputValue = getModelOutputValue(mlOutput, modelOutputFieldName, ignoreMissing, fullResponsePath);
+                        requestContext.setAttribute(newQueryField, modelOutputValue);
+
+                        if (newQueryField.startsWith("$.ext.") || newQueryField.startsWith("ext.")) {
```
could you please add doc when we will see `$.ext.` vs `ext.`?

mingshl replied:
sure

dhrubo-os commented on plugin/src/main/java/org/opensearch/ml/processor/ModelExecutor.java:
```diff
@@ -334,4 +342,29 @@ default List<String> writeNewDotPathForNestedObject(Object json, String dotPath)
         return path.replaceAll("\\[(\\d+)\\]", "$1\\.").replaceAll("\\['(.*?)']", "$1\\.").replaceAll("^\\$", "").replaceAll("\\.$", "");
     }
+
+    default void setRequestContextFromExt(SearchRequest request, PipelineProcessingContext requestContext) {
```
Can we please add doc for this method?

mingshl replied:
added

mingshl commented on plugin/src/main/java/org/opensearch/ml/searchext/MLInferenceRequestParametersUtil.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class MLInferenceRequestParametersUtil {
+
+    public static MLInferenceRequestParameters getMLInferenceRequestParameters(SearchRequest searchRequest) {
```
added

mingshl commented:
> Looks good overall, but can you add REST test cases with some non trivial examples?

I added two REST tests to verify in the new commit [add REST test and javadoc](https://github.com/opensearch-project/ml-commons/pull/3284/commits/0960694a1862c643ea52bb72a257b35481bf5d1e) and just rebased the changes from main branch

mingshl commented:
backport conflict due to this PR also changes in StringUtils, but it merge in main not backport to 2.x yet https://github.com/opensearch-project/ml-commons/pull/3342
